cmake/modules/packaging.cmake

```diff
@@ -91,16 +90,45 @@ macro(CREATE_PYTHON_EXE EXE_NAME MAIN_MODULE)
         "if __name__ == '__main__':\n"
         "    sys.exit(main())\n")
 
-    set(ZIP_STR 
-        "zip -qr ${CMAKE_CURRENT_BINARY_DIR}/${EXE_NAME}.zip ${PYTHON_SRC}")   
-    set(ZIP_MAIN_STR
-        "zip -qj ${CMAKE_CURRENT_BINARY_DIR}/${EXE_NAME}.zip ${BUILD_DIR_MAIN}/__main__.py")
-    set(ECHO_STR "echo '#!/usr/bin/env python ' | 
-        cat - ${CMAKE_CURRENT_BINARY_DIR}/${EXE_NAME}.zip > ${PACKAGER_BIN}")
-       
-    execute_process(COMMAND sh -c "${ZIP_STR}; ${ZIP_MAIN_STR}"
-              WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
-
-    execute_process(COMMAND sh -c "${ECHO_STR}; chmod a+x ${PACKAGER_BIN}")
+    # Generate a Python script to zip the sources.
+    #   *** We could use writepy() for Python files, but this introduces the
+    #   *** potential for compatibility problems, especially in RPMs.
+    file(WRITE "${BUILD_DIR_MAIN}/do_zip.py"
+        "import os\n"
+        "import stat\n"
+        "import zipfile\n"
+        "from io import BytesIO\n"
+        "\n"
+        "def addfile(fn, z):\n"
+        "  z.write(fn)\n"
+        "\n"
+        "# Write to a buffer so that the shebang can be prepended easily\n"
+        "wr_buf = BytesIO()\n"
+        "wr_buf.write('#!/usr/bin/env python' + os.linesep)\n"
+        "\n"
+        "z = zipfile.PyZipFile(wr_buf, 'w')\n")
+
+    foreach(PYFILE ${ARGN})
+        file(APPEND "${BUILD_DIR_MAIN}/do_zip.py"
+             "addfile('${PYFILE}', z)\n")
+    endforeach(PYFILE)
+
+    file(APPEND "${BUILD_DIR_MAIN}/do_zip.py"
+        "z.write('${BUILD_DIR_MAIN}/__main__.py', '__main__.py')\n"
+        "z.close()\n"
+        "\n"
+        "# Write out the buffer\n"
+        "with open('${PACKAGER_BIN}', 'wb') as f:\n"
+        "  f.write(wr_buf.getvalue())\n"
+        "  # Mark the file executable\n"
+        "  mode = os.fstat(f.fileno()).st_mode\n"
+        "  mode |= stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH\n"
+        "  os.fchmod(f.fileno(), stat.S_IMODE(mode))\n"
+        "\n"
+        "f.close()\n")
+
+    # Run Python to generate the zipped file
+    execute_process(COMMAND python "${BUILD_DIR_MAIN}/do_zip.py"
```

Comment:
Should we call `find_package(PythonInterp)` somewhere and use its cmake vars for pointing to the Python interpreter?
This will allow us to specify versions of Python (if we need to) and handle cases where Python isn't found.

Reply:
Good plan. There's some refactoring that could be done, too. The macro I changed here should probably be moved to a different file. E.g., FindPythonPackage.cmake in the same directory also invokes Python and may be a better home than packaging for the Python zip script.

Reply:
Eventually, I think we should use Python's setuptools to make wheel packages.
See: https://packaging.python.org/guides/tool-recommendations/